apollo-router/src/plugins/authentication/jwks.rs

```diff
@@ -256,3 +273,419 @@ impl Iterator for Iter<'_> {
+                    (Algorithm::ES384, AlgorithmParameters::EllipticCurve(params)) => {
+                        if params.curve == EllipticCurve::P384 {
+                            key.common.key_algorithm = Some(convert_algorithm(criteria.alg));
+                        }
+                    }
+                    (Algorithm::EdDSA, AlgorithmParameters::EllipticCurve(params)) => {
+                        if params.curve == EllipticCurve::Ed25519 {
+                            key.common.key_algorithm = Some(convert_algorithm(criteria.alg));
+                        }
+                    }
+                    _ => {
+                        // We'll ignore combinations we don't recognise
+                        continue;
+                    }
+                },
+            };
+
+            // If we get here we have a key that:
+            //  - may be used for signature verification
+            //  - has a matching algorithm, or if JWT has no algorithm, a matching key type
+            // It may have a matching kid if the JWT has a kid and it matches the key kid
+            //
+            // Multiple keys may meet the matching criteria, but they have a score. They get 1
+            // point for having an explicitly matching algorithm and 1 point for an explicitly
+            // matching kid. We will sort our candidates and pick the key with the highest score.
+
+            // If we find a key with a HIGHEST_SCORE, we will filter the list to only keep those
+            // with that score
+            if key_score == HIGHEST_SCORE {
+                found_highest_score = true;
+            }
+
+            candidates.push((key_score, (issuer.clone(), key)));
+        }
+    }
+
+    tracing::debug!(
+        "jwk candidates: {:?}",
+        candidates
+            .iter()
+            .map(|(score, (_, candidate))| (
+                score,
+                &candidate.common.key_id,
+                candidate.common.key_algorithm
+            ))
+            .collect::<Vec<(&usize, &Option<String>, Option<KeyAlgorithm>)>>()
+    );
+
+    if candidates.is_empty() {
+        None
+    } else {
+        // Only sort if we need to
+        if candidates.len() > 1 {
+            candidates.sort_by(|a, b| a.0.cmp(&b.0));
+        }
+
+        if found_highest_score {
+            Some(
+                candidates
+                    .into_iter()
+                    .filter_map(|(score, candidate)| {
+                        if score == HIGHEST_SCORE {
+                            Some(candidate)
+                        } else {
+                            None
+                        }
+                    })
+                    .collect(),
+            )
+        } else {
+            Some(
+                candidates
+                    .into_iter()
+                    .map(|(_score, candidate)| candidate)
+                    .collect(),
+            )
+        }
+    }
+}
+
+pub(super) fn extract_jwt<'a, 'b: 'a>(
```

Comment:
I think it would be useful to document what's the meaning of the return type here. Like what does that mean when it's `None` or `Some(Err)`  and so one

Reply:
Oh ok, sorry. Unfortunately it hasn't be done before :p now it's like dark magic